### DISCUSSION TITLE
Python Generator

### DISCUSSION THREAD

**Discussion by 32844583** [1101-GS3073-C/Python-Community-GS3073-C#545] [Ideas]
### Generator
**Generator 經常被使用，因為它們比 List 更節省內存，List 的所有值都會立即存儲在內存中。 當你使用 Generator 時，每次只會產生一個值(即使在 while 中執行 yield 也固定執行一次)，並且 Generator 一次只能在內存中保存一個值。 當您處理非常大的數據集且內存不夠時很有用。**

範例:
以下第一種用 list 實作產生大量集合元素的方法。可以看到裡面有一個區域變數 result_list，會一直將產生的字串加入到該 list 中，加完之後回傳整個結果 list。

![螢幕擷取畫面 2022-12-28 201202](https://user-images.githubusercontent.com/66071134/209810422-ea4f376f-93f7-40ec-b402-0d2f215bf796.jpg)

很明顯的這段程式碼最大的問題就是當指定的 num很大時，result_list 每次都儲存的結果會吃掉大量的記憶體資源，後果極可能造成程式 crash 掉。所以generator 就派上用場了，改寫程式如下。

![螢幕擷取畫面 2022-12-28 201213](https://user-images.githubusercontent.com/66071134/209810424-2ad5ba66-599a-4151-aa72-4ae16a52355c.jpg)

由於 generator 在迭代階段才運行程式邏輯的特性，所以根本不需要額外的區域變數來儲存每次迭代運算的結果，大大的降低記憶體資源的使用量，同時也可以達到與 list 方法一樣的結果。

https://mark1002.github.io/2018/07/31/python-generator-學習紀錄/

**Comment by williamyen0623** [comment]
似乎在大型計畫、資料量巨大的時候才會需要用到，畢竟現在的電腦性能肯定不管寫什麼都有辦法handle，不過學起來也有利無弊，
說不定將來求職會用到？

**Comment by HsuChen88** [comment]
generator在時間和記憶體上的效率都比list更好
